File: LAB_8/Builder.java
class Pizza {
    String dough;
    String sauce;
    String topping;


    private Pizza(PizzaBuilder builder) {
        this.dough = builder.dough;
        this.sauce = builder.sauce;
        this.topping = builder.topping;
    }

    public static class PizzaBuilder {
        String dough;
        String sauce;
        String topping;

        public PizzaBuilder dough(String dough) {
            this.dough = dough;
            return this;
        }

        public PizzaBuilder sauce(String sauce) {
            this.sauce = sauce;
            return this;
        }

        public PizzaBuilder topping(String topping) {
            this.topping = topping;
            return this;
        }

        Pizza build() {
            return new Pizza(this);
        }
    }

    public void display() {
        System.out.println("\\/ Pizza ingredients below\\/");
        System.out.println("Dough: " + dough);
        System.out.println("Sauce: " + sauce);
        System.out.println("Topping: " + topping);
    }

    public static void main(String[] args) {
        Pizza pizza = new Pizza.PizzaBuilder()
                .dough("dough")
                .sauce("tomato")
                .topping("cheese + ham")
                .build();

        pizza.display();
    }
}

File: LAB_8/Builder.cpp
#include <iostream>
#include <string>

class Pizza {
    private:
        std::string dough;
        std::string sauce;
        std::string topping;

    public:
        Pizza(const std::string& dough, const std::string& sauce, const std::string& topping) : dough(dough), sauce(sauce), topping(topping) {}

        class Builder {
            private:
                std::string dough;
                std::string sauce;
                std::string topping;
    
            public:
                Builder& setDough(std::string dough) {
                    this->dough = dough;
                    return *this;
                }

                Builder& setSauce(std::string sauce) {
                    this->sauce = sauce;
                    return *this;
                }

                Builder& setTopping(std::string topping) {
                    this->topping = topping;
                    return *this;
                }

                Pizza build() {
                    return Pizza(dough, sauce, topping);
                }
            };

        
        void display() const {
            std::cout << "\\/ Pizza ingredients below \\/" << std::endl;
            std::cout << "Dough: " << dough << std::endl;
            std::cout << "Sauce: " << sauce << std::endl;
            std::cout << "Topping: " << topping << std::endl;
            
        }
};

int main() {
    Pizza pizza = Pizza::Builder()
                    .setDough("Thin")
                    .setSauce("Tomato")
                    .setTopping("Cheese + ham")
                    .build();
    
    pizza.display();
    
    return 0;
}